From b4c17cf7b515a5d7e27e40004ed81137647ce12a Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Wed, 30 Nov 2022 16:20:29 +0100
Subject: [PATCH] ARM: dts: stm32: add cpufreq support on stm32mp15

This commit adds cpufreq support on stm32mp15 SOC. STM32 cpufreq uses
operating points V2 bindings (no legacy). Nvmem cells have to be used to
know the chip version and then which OPPs are available. Note that STM32
cpufreq driver is mainly based on "cpufreq-dt" driver.
The 800Mhz opp is allowed until temperature reaches 95 degrees.

Change-Id: Idf4c1c58f8c22a235aaa53e1e664e7f52814ab74
Signed-off-by: Pascal Paillet <p.paillet@st.com>
---
 arch/arm/boot/dts/st/stm32mp151.dtsi      | 11 ++++++-
 arch/arm/boot/dts/st/stm32mp153.dtsi      |  3 +-
 arch/arm/boot/dts/st/stm32mp157a-ed1.dts  |  8 +++++
 arch/arm/boot/dts/st/stm32mp157c-ed1.dts  |  8 +++++
 arch/arm/boot/dts/st/stm32mp157d-ed1.dts  |  8 +++++
 arch/arm/boot/dts/st/stm32mp157f-ed1.dts  |  8 +++++
 arch/arm/boot/dts/st/stm32mp15xa.dtsi     |  8 +++++
 arch/arm/boot/dts/st/stm32mp15xd.dtsi     | 37 +++++++++++++++++++++++
 arch/arm/boot/dts/st/stm32mp15xx-dkx.dtsi |  8 +++++
 9 files changed, 97 insertions(+), 2 deletions(-)

--- a/arch/arm/boot/dts/st/stm32mp151.dtsi
+++ b/arch/arm/boot/dts/st/stm32mp151.dtsi
@@ -18,12 +18,21 @@
 
 		cpu0: cpu@0 {
 			compatible = "arm,cortex-a7";
-			clock-frequency = <650000000>;
 			device_type = "cpu";
 			reg = <0>;
+			clock-names = "cpu";
+			operating-points-v2 = <&cpu0_opp_table>;
+			nvmem-cells = <&part_number_otp>;
+			nvmem-cell-names = "part_number";
+			#cooling-cells = <2>;
 		};
 	};
 
+	cpu0_opp_table: cpu0-opp-table {
+		compatible = "operating-points-v2";
+		opp-shared;
+	};
+
 	arm-pmu {
 		compatible = "arm,cortex-a7-pmu";
 		interrupts = <GIC_SPI 200 IRQ_TYPE_LEVEL_HIGH>;
--- a/arch/arm/boot/dts/st/stm32mp153.dtsi
+++ b/arch/arm/boot/dts/st/stm32mp153.dtsi
@@ -10,9 +10,10 @@
 	cpus {
 		cpu1: cpu@1 {
 			compatible = "arm,cortex-a7";
-			clock-frequency = <650000000>;
 			device_type = "cpu";
 			reg = <1>;
+			clock-names = "cpu";
+			operating-points-v2 = <&cpu0_opp_table>;
 		};
 	};
 
--- a/arch/arm/boot/dts/st/stm32mp157a-ed1.dts
+++ b/arch/arm/boot/dts/st/stm32mp157a-ed1.dts
@@ -136,6 +136,14 @@
 	status = "okay";
 };
 
+&cpu0 {
+	cpu-supply = <&vddcore>;
+};
+
+&cpu1 {
+	cpu-supply = <&vddcore>;
+};
+
 &crc1 {
 	status = "okay";
 };
--- a/arch/arm/boot/dts/st/stm32mp157c-ed1.dts
+++ b/arch/arm/boot/dts/st/stm32mp157c-ed1.dts
@@ -136,6 +136,14 @@
 	status = "okay";
 };
 
+&cpu0 {
+	cpu-supply = <&vddcore>;
+};
+
+&cpu1 {
+	cpu-supply = <&vddcore>;
+};
+
 &crc1 {
 	status = "okay";
 };
--- a/arch/arm/boot/dts/st/stm32mp157d-ed1.dts
+++ b/arch/arm/boot/dts/st/stm32mp157d-ed1.dts
@@ -136,6 +136,14 @@
 	status = "okay";
 };
 
+&cpu0 {
+	cpu-supply = <&vddcore>;
+};
+
+&cpu1 {
+	cpu-supply = <&vddcore>;
+};
+
 &crc1 {
 	status = "okay";
 };
--- a/arch/arm/boot/dts/st/stm32mp157f-ed1.dts
+++ b/arch/arm/boot/dts/st/stm32mp157f-ed1.dts
@@ -136,6 +136,14 @@
 	status = "okay";
 };
 
+&cpu0 {
+	cpu-supply = <&vddcore>;
+};
+
+&cpu1 {
+	cpu-supply = <&vddcore>;
+};
+
 &crc1 {
 	status = "okay";
 };
--- a/arch/arm/boot/dts/st/stm32mp15xa.dtsi
+++ b/arch/arm/boot/dts/st/stm32mp15xa.dtsi
@@ -3,3 +3,11 @@
  * Copyright (C) STMicroelectronics 2023 - All Rights Reserved
  * Author: Alexandre Torgue <alexandre.torgue@foss.st.com> for STMicroelectronics.
  */
+
+&cpu0_opp_table {
+		opp-650000000 {
+			opp-hz = /bits/ 64 <650000000>;
+			opp-microvolt = <1200000>;
+			opp-supported-hw = <0x1>;
+		};
+};
--- a/arch/arm/boot/dts/st/stm32mp15xd.dtsi
+++ b/arch/arm/boot/dts/st/stm32mp15xd.dtsi
@@ -3,3 +3,40 @@
  * Copyright (C) STMicroelectronics 2023 - All Rights Reserved
  * Author: Alexandre Torgue <alexandre.torgue@foss.st.com> for STMicroelectronics.
  */
+
+&cpu0_opp_table {
+		opp-800000000 {
+			opp-hz = /bits/ 64 <800000000>;
+			opp-microvolt = <1350000>;
+			opp-supported-hw = <0x2>;
+		};
+		opp-400000000 {
+			opp-hz = /bits/ 64 <400000000>;
+			opp-microvolt = <1200000>;
+			opp-supported-hw = <0x2>;
+			opp-suspend;
+		};
+};
+
+&cpu_thermal {
+	trips {
+		cpu-crit {
+			temperature = <105000>;
+			hysteresis = <0>;
+			type = "critical";
+		};
+
+		cpu_alert: cpu-alert {
+			temperature = <95000>;
+			hysteresis = <10000>;
+			type = "passive";
+		};
+	};
+
+	cooling-maps {
+		map0 {
+			trip = <&cpu_alert>;
+			cooling-device = <&cpu0 1 1>;
+		};
+	};
+};
--- a/arch/arm/boot/dts/st/stm32mp15xx-dkx.dtsi
+++ b/arch/arm/boot/dts/st/stm32mp15xx-dkx.dtsi
@@ -141,6 +141,14 @@
 	status = "okay";
 };
 
+&cpu0 {
+	cpu-supply = <&vddcore>;
+};
+
+&cpu1 {
+	cpu-supply = <&vddcore>;
+};
+
 &crc1 {
 	status = "okay";
 };
